### PR TITLE
Make IntSet a generic ordinal set PackedSet[A] + new module packedsets

### PR DESCRIPTION
**Pull request by landerlo**
This makes intsets generic for ordinal types like enums or distinct of ints. For cases when multiple intsets with complex logic making them generic boosts their usability and type safety.
E.g. iqqn my use case I need to distinguish different ids making them distinct and get compilation errors if
I do any set operations with incompatible ids.

The PR achieves this maintaining full backwards compatibility and same performance.
As a drop in replacement, the current api is unchanged and IntSet is aliased to OrdSet[int].
OrdSet is for ordinal set. I realise there might be a confusion with ordered sets, so naming suggestions welcome.

Better error messages on compilation errors will be achieved when the concepts are ready, but I didn't want to tie it to a future roadmap, and the current solution achieves the stated goals satisfactorily IMHO.

### CODE REVIEW

File: lib/pure/collections/intsets.nim
Comment (timotheecour) [inline lib/pure/collections/intsets.nim]: shouldn't it be: `OrdSet*[A: Ordinal]` or something similar?

File: lib/pure/collections/intsets.nim
Comment (timotheecour) [inline lib/pure/collections/intsets.nim]: where is `A` used in type fields?

File: lib/pure/collections/intsets.nim
Comment (timotheecour) [inline lib/pure/collections/intsets.nim]: not `since`, `toIntSet` symbol already existed prior your PR (albeit with different implementation)

ditto wherever relevant

File: lib/pure/collections/intsets.nim
Comment (timotheecour) [inline lib/pure/collections/intsets.nim]: what about instead adding `lib/pure/collections/ordsets.nim` (with no mention of intsets there) and then making  `lib/pure/collections/intsets.nim` a thin wrapper over `OrdSet[int]` that imports ordsets ?

File: lib/pure/collections/intsets.nim
Comment (landerlo) [reply lib/pure/collections/intsets.nim]: Thank you, fixed

File: lib/pure/collections/intsets.nim
Comment (landerlo) [reply lib/pure/collections/intsets.nim]: I kept what it was in devel, toIntSet already has since 1.3

File: lib/pure/collections/intsets.nim
Comment (timotheecour) [reply lib/pure/collections/intsets.nim]: if an item is fixed (as opposed to pushback), you can click "resolve conversation" ;-)

File: lib/pure/collections/intsets.nim
Comment (landerlo) [reply lib/pure/collections/intsets.nim]: I thought about that, but I wanted to keep it as simple as possible to spot the differences, does github detect moved to show the diff?

Comment (landerlo) [reply lib/pure/collections/intsets.nim]: Now that I think of it, with the move in a separate commit the diff is still seen in the first commit, so I did the move to ordsets and kept intsets as an alias to ordsets. Not sure if that causes any issue with backwards compatibility. This is my first nim code,

Comment (landerlo) [reply lib/pure/collections/intsets.nim]: Still, there are lots of imports in the compiler. is there any way of importing transitively in nim? i.e. I leave intset just with the IntSet type alias and IntSet specific procs and import from ord. It seems that if some old user code imports intset it won't get the other procs transitively even if intset imports from ordset, is this correct?

**Comment by disruptek** [comment]
I liked it better before it touched the compiler, but whatever.  I think we have to wait until `Ordinal` is a concept before it will work for `distinct` types.

File: lib/pure/collections/intsets.nim
Comment (timotheecour) [reply lib/pure/collections/intsets.nim]: did you try this?
```nim
# intsets.nim
import std/ordsets
export ordsets
```

**Comment by timotheecour** [comment]
> I liked it better before it touched the compiler, but whatever. 

/cc @landerl
I agree with @disruptek , I don't think there's a need to change client code, see my answer here: https://github.com/nim-lang/Nim/pull/15564#discussion_r504416600


> I think we have to wait until Ordinal is a concept before it will work for distinct types.

what do you mean? this passes:
```nim
type int2 = distinct int
doAssert int2 is Ordinal
```

File: lib/pure/collections/intsets.nim
Comment (landerlo) [reply lib/pure/collections/intsets.nim]: Ok, I didn't know about exports. However it says it 'Error: cannot export: ordsets'

**Comment by landerlo** [comment]
> > I liked it better before it touched the compiler, but whatever.
> 
> /cc @landerl
> I agree with @disruptek , I don't think there's a need to change client code, see my answer here: [#15564 (comment)](https://github.com/nim-lang/Nim/pull/15564#discussion_r504416600)
> 
> > I think we have to wait until Ordinal is a concept before it will work for distinct types.
> 
> what do you mean? this passes:
> 
> ```nim
> type int2 = distinct int
> doAssert int2 is Ordinal
> ```
See my comment above, making all the type params A: Ordinal didn't work for me, but just making OrdSet[A: Ordinal] and the api procs [A] seems to work and has now better compilation errors, so yes, concepts are not necessary for this.

File: lib/pure/collections/intsets.nim
Comment (timotheecour) [reply lib/pure/collections/intsets.nim]: why? do you have a minimal example (or simply a temporary branch in your nim fork ) where it gives this error so we can look into it?
for eg this works and is a common pattern:
```nim
# in foo.nim
import std/tables
export tables
```

File: lib/pure/collections/intsets.nim
Comment (landerlo) [reply lib/pure/collections/intsets.nim]: I hadn't been able to make [A: Ordinal] work and now I see that apparently you cannot have both the type and the procs in the api constrained to ordinal. Nevertheless unconstraining the Api and making the type constrained works and it's better for the type safety. Changed that.

**Comment by disruptek** [comment]
I mean that `Ordinal` doesn’t capture a type than can nominally behave as an ordinal because it’s magic and not a concept.

**Comment by landerlo** [comment]
> I mean that `Ordinal` doesn’t capture a type than can nominally behave as an ordinal because it’s magic and not a concept.

At the moment is magic but there could be a concept modeling Ordinal  with `ord(a: A): A`, `fromOrd(o: int): A`, and an implicit assumption that the constraint: `forall x: int ord(fromOrd(int)) == x && forall a: A fromOrd(ord(a)) == a`, that's left to an axiom for ints, enums and distincts of an ordinal. But maybe user could define other ordinal types to benefit from things like ordsets.

That Ordinal concept could be further generalised to Isomorphic. so concept `Ordinal[A] = Isomorphic[A, int]` with the above signature.

File: tests/stdlib/tordsets.nim
Comment (timotheecour) [inline tests/stdlib/tordsets.nim]: new modules should not be importable without `std/` prefix to avoid name clashes with future nimble packages; maybe ordsets should be under `std/ordsets` even though it's a collection

File: lib/pure/collections/ordsets.nim
Comment (timotheecour) [inline lib/pure/collections/ordsets.nim]: is this used at all here? why not move to intsets.nim
(make sure tordsets.nim sill works, if needed you can also import std/intsets from there)

File: tests/stdlib/tordsets.nim
Comment (landerlo) [reply tests/stdlib/tordsets.nim:1]: What does this entail? moving the ordsets.nim to Nim/lib/std/ folder?

Comment (timotheecour) [reply tests/stdlib/tordsets.nim]: it's marked as resolved but was it resolved? ;-)

Comment (timotheecour) [reply tests/stdlib/tordsets.nim]: nevermind, i hadn't clicked refresh!

File: tests/stdlib/tordsets.nim
Comment (timotheecour) [inline tests/stdlib/tordsets.nim]: you've changed this to `import std/ordsets` (good) but the problem remains:
`import ordsets` should not compile IMO, there's a policy (i forgot where) that new modules show require std prefix; that's to avoid clashes with nimble packages

hence my last remark.
can you move ordsets to: `lib/std/ordsets` ?

yes I know it's a collection but that's a less important point than module name global scope pollution IMO

feel free to check with @araq before doing this change

Comment (landerlo) [reply tests/stdlib/tordsets.nim]: I've pushed the move to lib/std, I'm switching off now, but I can check with him tomorrow. Thank you for the review, this is literally my first nim code so I'm learning along.

Comment (landerlo) [reply tests/stdlib/tordsets.nim]: I've checked RFCs about nimble and naming and other that the std/ prefix as you suggested I couldn't find any clear criteria about whether it should go on folder lib/std or lib/pure/collections. What's your executive decision on this?
I've amended the commits to be clearer, 1. Full change over intset file for an easier diff of the change. 2. Rename file to ordsets and factor out intsets with the ordsets exports. 3. Just the tentative file move to to /lib/std

Comment (timotheecour) [reply tests/stdlib/tordsets.nim]: see https://github.com/nim-lang/Nim/pull/8024#issuecomment-396943421
> There is no reason whatsoever we can't have a long deprecation period for import strutils and the like, we can even decide the old stdlib modules should be seen as "keyword-like" and do not have to deprecate it at all (!) but all the new stuff ends up in std.

I'm adding this convention in this PR: https://github.com/nim-lang/Nim/pull/15590

Comment (landerlo) [reply tests/stdlib/tordsets.nim]: The last commit moves `ordsets.nim` to `lib/std`. Is there anything else that needs doing?

**Comment by landerlo** [comment]
@timotheecour is there anything else needed for the PR? ordsets is in lib/std folder as per your request.

**Comment by disruptek** [comment]
I feel bad for how a simple PR became such a large bikeshed.  Can this go in?  Please?

File: tests/stdlib/tordsets.nim
Comment (timotheecour) [inline tests/stdlib/tordsets.nim]: nit: why the quoting of A?
ditto elsewhere

File: tests/stdlib/tordsets.nim
Comment (timotheecour) [inline tests/stdlib/tordsets.nim]: instead of "should fail" (etc), test for it using:
`doAssert not compiles(union(distT, initOrdSet[int]()))` or similar

ditto elsewhere

File: tests/stdlib/tordsets.nim
Comment (timotheecour) [inline tests/stdlib/tordsets.nim]: style: user types start with capital

ditto elsewhere eg `type id = distinct int`

File: tests/stdlib/tordsets.nim
Comment (timotheecour) [inline tests/stdlib/tordsets.nim]: use `doAssert` in tests, not `assert`

File: lib/std/ordsets.nim
Comment (timotheecour) [inline lib/std/ordsets.nim]: (could be done in cleanup post-merge) tests must be merged with tests/stdlib/tordsets.nim, not  in a isMainModule block; and possibly reduced in case some of it is redundant with tests/stdlib/tordsets.nim

File: lib/std/ordsets.nim
Comment (timotheecour) [inline lib/std/ordsets.nim]: this is the only real blocker.
```nim
proc main() =
  type Foo = enum k0 k1 k2
  var a = initOrdSet[Foo]()
  a.incl k0
  a.incl k2
  echo $a # got {0, 2}, expected {k0, k2}
main()
```

File: lib/std/ordsets.nim
Comment (timotheecour) [inline lib/std/ordsets.nim]: true but incomplete (excludes char, bool etc).
instead, maybe just say: Supports any `Ordinal` type.

**Comment by landerlo** [comment]
> I feel bad for how a simple PR became such a large bikeshed. Can this go in? Please?

It replaces a collection used in the compiler and I imagine quite popular, so I'm glad it's been properly scrutinized, but I think after the iterations it's ready to be merged. I also checked there is no performance overhead so the zero cost abstraction works as expected.

File: lib/std/ordsets.nim
Comment (timotheecour) [inline lib/std/ordsets.nim]: nit (can be made post merge in a cleanup PR) all these private procs could be renamed eg `ordSetGet` etc

**Review by timotheecour — APPROVED** [review]
LGTM, fixing `$` is the main thing missing, the rest are nits that can be addressed in future cleanup PRs

note to other reviewers:
github and git diff aren't smart enough here, to view the diff use:
```
git diff 675cf57:lib/pure/collections/intsets.nim lib/std/ordsets.nim
git diff 675cf57:tests/stdlib/tintsets.nim tests/stdlib/tordsets.nim
```
(adapt `675cf57` as needed)

File: lib/std/ordsets.nim
Comment (landerlo) [reply lib/std/ordsets.nim]: So the change is to supports distincts with any ordinal base? I didn't check any other cases other than int but that makes sense.

Comment (timotheecour) [reply lib/std/ordsets.nim]: I mean that `Ordinal` is already defined elsewhere so there's no need to specify again what it means (and excluding some members from its definition eg `char`, `bool` etc). I just checked locally that `bool` and `enum` do work with ordsets.

File: lib/std/ordsets.nim
Comment (landerlo) [reply lib/std/ordsets.nim]: There is an ordSetGet already, the old method is kept as it seems it's necessary to keep compatibility with tests for version 1.0, tests would fail otherwise If I remember correctly, but I can double-check.

**Comment by landerlo** [comment]
> LGTM, fixing `$` is the main thing missing, the rest are nits that can be addressed in future cleanup PRs
> 
> note to other reviewers:
> github and git diff aren't smart enough here, to view the diff use:
> 
> ```
> git diff 675cf57:lib/pure/collections/intsets.nim lib/std/ordsets.nim
> git diff 675cf57:tests/stdlib/tintsets.nim tests/stdlib/tordsets.nim
> ```
> 
> (adapt `675cf57` as needed)

All the changes are in the first commit to be more visible. Then there is one renaming it.

File: lib/std/ordsets.nim
Comment (landerlo) [reply lib/std/ordsets.nim]: Ok, so implementing proc $, is Nim's string representation? I'll change that to get the ordinal name for the ints we store in the bitset

File: tests/stdlib/tordsets.nim
Comment (landerlo) [reply tests/stdlib/tordsets.nim]: It's a typedesc arg for the template, isn't the quoting required for the typecast?

File: lib/std/ordsets.nim
Comment (landerlo) [reply lib/std/ordsets.nim]: I carried over the main tests existing from intset but I'll migrate to tordset.nim if there is anything not covered already in those tests

File: tests/stdlib/tordsets.nim
Comment (timotheecour) [reply tests/stdlib/tordsets.nim]: no, this works:
```nim
  template fn(A: typedesc) = echo A(1)
  fn(float)
```

File: lib/std/ordsets.nim
Comment (timotheecour) [reply lib/std/ordsets.nim]: all that's needed is:
```nim
# in dollarImpl
result.add($(ord(key)))
=>
result.add $key
```

note for distinct:
```nim
when true:
  import std/ordsets
  type Foo = distinct int16
  proc `$`(a: Foo): string {.borrow.} # `echo a` below won't work without `$` defined, as expected
  var a = initOrdSet[Foo]()
  a.incl Foo(3)
  a.incl Foo(4)
  echo a
```

File: lib/std/ordsets.nim
Comment (timotheecour) [reply lib/std/ordsets.nim]: > There is an ordSetGet already

I don't see it?

> keep compatibility with tests for version 1.0

that doesn't make sense, this is private proc

in any case, feel free to delay this to a future cleanup PR since it's not client facing

File: lib/pure/collections/intsets.nim
Comment (timotheecour) [inline lib/pure/collections/intsets.nim]: `(1,5,1)` now

File: lib/std/ordsets.nim
Comment (landerlo) [reply lib/std/ordsets.nim]: My mistake I was thinking about the public methods toIntSet/toOrdSet. I'll rename intSetGet

File: lib/pure/collections/intsets.nim
Comment (landerlo) [reply lib/pure/collections/intsets.nim]: toIntset was public and had the since 1.3 from before, so shouldn't we keep that version info?

Comment (timotheecour) [reply lib/pure/collections/intsets.nim]: misread

**Review by timotheecour — APPROVED** [review]

File: lib/pure/collections/intsets.nim
Comment (timotheecour) [inline lib/pure/collections/intsets.nim]: CI fails:

Error: undeclared field: 'id' for type system.int [type declared in /Users/runner/work/Nim/Nim/lib/system/basic_types.nim(2, 3)]

**Comment by timotheecour** [comment]
I'd like to merge this soon unless there are further objections
/cc @Araq

**Comment by Araq** [comment]
Sorry I'm late to the party, but is ``OrdSet`` really the best name we could come up with? How about a ``BitmaskSet``, ``BitSet``, ``HashBitSet``?

**Comment by landerlo** [comment]
> Sorry I'm late to the party, but is `OrdSet` really the best name we could come up with? How about a `BitmaskSet`, `BitSet`, `HashBitSet`?

There is the `IntSet` alias, which is basically what people would expect when they read `BitSet`. But if we underlie that it's about Ordinals, this showcases Nim's generics and Ordinals features.
In my use case I have sets of different id types (distincts of int). The generic Ordinal mask set gets you  type safety when you insert and combine different sets with the performance and memory footprint of a bitset. I cannot think of any other language I know where this can be provided this seamlessly.

That said, I have some doubts about `OrdSet` as the Ord could be thought as Ordered, rather than Ordinal. Please advice a name and I'll rename this.

**Comment by timotheecour** [comment]
`BitSet` (in `std/bitsets`) would be the best IMO (short and self documenting) but there's already `compiler/bitsets.nim`; this could be renamed to avoid confusion but with small caveat that compiler is a nimble package too so could potentially break client code relying on it (but compiler code should be changeable IMO, outside of a stable subset `compilerapi`).

Next best would be BitMaskSet (in std/bitmasksets) but it's rather lengthy

**Comment by Araq** [comment]
``BucketSet`` ? Sorry, I'm still looking for a good name. Or maybe ``ClusterSet`` as it's a set of bit clusters.

**Comment by timotheecour** [comment]
`PackedSet` or `SparseSet` also is possible; anything with 2 words is fine (`HashBitSet` is a bit long)

**Comment by disruptek** [comment]
`OrdinalSet`

**Comment by Araq** [comment]
Let's go with ``PackedSet``.

**Comment by timotheecour** [comment]
@landerlo ok so PackedSet + std/packedsets + tpackedsets.nim; hopefully it won't be too painful to do this renaming; thanks!

**Comment by landerlo** [comment]
No problem, renamed to PackedSet/packedsets

**Review by timotheecour — APPROVED** [review]

**Comment by timotheecour** [comment]
LGTM; good to merge?

**Comment by timotheecour** [comment]
> landerlo requested a review from timotheecour 16 minutes ago

I've already did a full review and gave my LGTM ...

@araq or other nim member can i merge this?

**Comment by Araq** [comment]
Sorry, I thought it already was merged.

**Comment by landerlo** [comment]
I do realise now that sed doesn't go through the commit messages, so the commit log still talks about OrdSet and not  the renamed PackedSet. Nevermind.

**Comment by disruptek** [comment]
Thanks for all your work on this, @landerlo -- I know it was a slog but this is a nice contribution. ❤️

**Comment by disruptek** [comment]
I hate to say it, but this is causing the `items` iterator against an `IntSet` to fail:

https://github.com/disruptek/gram/runs/1419861309?check_suite_focus=true

```nim
/home/adavidoff/nims/lib/std/packedsets.nim(370, 14) Error: type mismatch: got <PackedSet[system.int]>
but expected one of: 
iterator items(a: cstring): char
  first type mismatch at position: 1
  required type for a: cstring
  but expression 's' is of type: PackedSet[system.int]
iterator items(a: string): char
  first type mismatch at position: 1
  required type for a: string
  but expression 's' is of type: PackedSet[system.int]
iterator items(n: NimNode): NimNode
  first type mismatch at position: 1
  required type for n: NimNode
  but expression 's' is of type: PackedSet[system.int]
iterator items[IX, T](a: array[IX, T]): T
  first type mismatch at position: 1
  required type for a: array[IX, T]
  but expression 's' is of type: PackedSet[system.int]
iterator items[N, E; F: static[GraphFlags]](graph: Graph[N, E, F]): N
  first type mismatch at position: 1
  required type for graph: Graph[items.N, items.E, items.F]
  but expression 's' is of type: PackedSet[system.int]
iterator items[T: char](a: openArray[T]): T
  first type mismatch at position: 1
  required type for a: openArray[T: char]
  but expression 's' is of type: PackedSet[system.int]
iterator items[T: enum](E: typedesc[T]): T
  first type mismatch at position: 1
  required type for E: type T: enum
  but expression 's' is of type: PackedSet[system.int]
iterator items[T: not char](a: openArray[T]): lent2 T
  first type mismatch at position: 1
  required type for a: openArray[T: not char]
  but expression 's' is of type: PackedSet[system.int]
iterator items[T](a: seq[T]): lent2 T
  first type mismatch at position: 1
  required type for a: seq[T]
  but expression 's' is of type: PackedSet[system.int]
iterator items[T](a: set[T]): T
  first type mismatch at position: 1
  required type for a: set[T]
  but expression 's' is of type: PackedSet[system.int]
iterator items[T](s: HSlice[T, T]): T
  first type mismatch at position: 1
  required type for s: HSlice[items.T, items.T]
  but expression 's' is of type: PackedSet[system.int]

expression: items(s)
```

**Comment by timotheecour** [comment]
@disruptek can you provide a minimized reproducible example?

**Comment by disruptek** [comment]
Is gram too complex for you?

**Comment by timotheecour** [comment]
that's basically what everyone has to go through when submitting bug reports, providing minimized reproducing examples. More work for the reporter, less work for everyone else looking at it.

**Comment by disruptek** [comment]
Yeah, but only if they didn't manage to get their package into `important_packages` first.  If they were able to do that, then for some reason the developers are willing to do the same work themselves.  Weird.

**Comment by timotheecour** [comment]
* if its's a compiler bug: minimize with 0 stdlib module dependencies, when possible
* if it's a stdlib bug: minimize with as few stdlib module dependencies, ideally just the module in question
it's about minimizing total work = 1 * work(bug reporter) + N * work(people looking at issue)

**Comment by disruptek** [comment]
Let's agree that you'll be the `1` to do the work and all `N` of us will reap the benefits.

**Comment by Araq** [comment]
As a compromise, both modules could expand a common template implementation -- then there cannot be regressions like this.

**Comment by landerlo** [comment]
@disruptek Sorry for the regression. Tests test int iterators but there must be some issue why the iterator is not properly inferred.
I've created a PR that simply reverts IntSet to previous impl rather than as an alias for PackedSet[int], that should do the trick for now. I want to dig deeper into the issue anyway at some point.
Rollbak IntSet as before:
https://github.com/nim-lang/Nim/pull/16059

**Comment by disruptek** [comment]
Can't we just fix it instead?

**Comment by timotheecour** [comment]
> As a compromise, both modules could expand a common template implementation -- then there cannot be regressions like this.

=> https://github.com/nim-lang/Nim/pull/16060, much cleaner

**Comment by disruptek** [comment]
Nice work!  Another good reason to be explicit with iterators, at least until this is fixed.

**Pull request merged**
